src/noun-enjs.ts

```diff
@@ -0,0 +1,204 @@
+export type EnjsFunction = (n: Noun) => Json;
+type frondOpt = { tag: string; get: EnjsFunction };
+
+const frond = function (opts: frondOpt[]): EnjsFunction {
+  return function (noun) {
+    if (!(noun instanceof Cell && noun.head instanceof Atom)) {
+      throw new Error("frond: noun not cell with tag head");
+    }
+    const tag = Atom.cordToString(noun.head);
+    for (let i = 0; i < opts.length; i++) {
+      if (tag === opts[i].tag) {
+        return { [tag]: opts[i].get(noun.tail) };
+      }
+    }
+    throw new Error("frond: unknown tag" + tag);
+  };
+};
+
+const tuple = function(funs: EnjsFunction[]): EnjsFunction {
+  return function (noun) {
+    let o = [];
+    while (funs.length > 1) {
+      if (noun.isAtom()) {
+        throw new Error("tuple: noun too shallow");
+      }
+      o.push(funs[0](noun.head));
+      funs.splice(0, 1);
+      noun = noun.tail;
+    }
+    o.push(funs[0](noun));
+    return o;
+  }
+}
+
+type PairCell = { nom: string; get: EnjsFunction };
+const pairs = function (cels: PairCell[]): EnjsFunction {
+  return function (noun) {
+    let i = 0;
+    let o: Record<string, Json> = {};
+    while (i < cels.length - 1) {
+      if (!(noun instanceof Cell)) {
+        throw new Error("pairs: noun too shallow");
+      }
+      o[cels[i].nom] = cels[i].get(noun.head);
+      noun = noun.tail;
+      i++;
+    }
+    o[cels[i].nom] = cels[i].get(noun);
+    return o;
+  };
+};
+const pair = function (
+  na: string,
+  ga: EnjsFunction,
+  nb: string,
+  gb: EnjsFunction
+): EnjsFunction {
+  return pairs([
+    { nom: na, get: ga },
+    { nom: nb, get: gb },
+  ]);
+};
+
+const bucwut = function (opts: EnjsFunction[]): EnjsFunction {
+  return function (noun) {
+    for (let i = 0; i < opts.length; i++) {
+      try {
+        const res = opts[i](noun);
+        return res;
+      } catch (e) {
+        continue;
+      }
+    }
+    throw new Error("bucwut: no matches");
+  };
+};
+
+//  buccen: like frond, but without the wrapper object
+const buccen = function (opts: frondOpt[]): EnjsFunction {
+  return function (noun) {
+    if (!(noun instanceof Cell && noun.head instanceof Atom)) {
+      throw new Error("buccen: noun not cell with tag head");
+    }
+    const tag = Atom.cordToString(noun.head);
+    for (let i = 0; i < opts.length; i++) {
+      if (tag === opts[i].tag) {
+        return opts[i].get(noun.tail);
+      }
+    }
+    throw new Error("buccen: unknown tag" + tag);
+  };
+};
+
+const array = function (item: EnjsFunction): (n: Noun) => Json[] {
+  return function (noun) {
+    let a: Json[] = [];
+    while (noun instanceof Cell) {
+      a.push(item(noun.head));
+      noun = noun.tail;
+    }
+    return a;
+  };
+};
+
+const tree = function (item: EnjsFunction): (n: Noun) => Json[] {
+  return function (noun) {
+    let a: Json[] = [];
+    if (noun instanceof Cell) {
+      if (!(noun.tail instanceof Cell)) {
```

Comment:
is this guaranteed to be the case at every level of the tree?

Reply:
Yes, see also the definition for `+tree` in hoon.hoon. Every node is either `~` or a triple (with `n`ode value, `l`eft & `r`ight branches.)